### PR TITLE
Add locale support to WebChat

### LINKED ISSUE #3296: Support locale at StartConversation
##Issue
Original issue: microsoft/botframework-obi#75

This issue tracks a frequently requested feature to allow a user locale when starting new conversations. The user locale is needed so that bots can send a properly localized Greeting Message to users.

##Proposed change
 WebChat - update the protocol to support locale and userId aligned with DirectLineJS

Tracking DirectLineJS issue: https://github.com/microsoft/BotFramework-DirectLineJS/issues/292

### PR DESCRIPTION
**Pull request by timenick**
Fixes #3296 

## Description
Update createDirectLine to accept and pass locale option.

### CODE REVIEW

File: __tests__/html/simple.localeSupport.html [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+              conversationStartLocale: locale
+            }),
+            store: createStore()
+          },
+          document.getElementById(frame)
+        );
+      }
+
+      (async function() {
+        await conversationStart('webchat1');
```
Comment (compulim) [inline __tests__/html/simple.localeSupport.html:40]: Please split into 5 tests.

File: packages/bundle/src/createDirectLine.js [hunk 2]
```diff
@@ -10,7 +10,8 @@ export default function createDirectLine({
   streamUrl,
   token,
   watermark,
-  webSocket
+  webSocket,
+  conversationStartLocale
```
Comment (compulim) [inline packages/bundle/src/createDirectLine.js:14]: Sort.

File: packages/bundle/src/createDirectLine.js [hunk 3]
```diff
@@ -23,6 +24,7 @@ export default function createDirectLine({
     token,
     watermark,
     webSocket,
+    conversationStartLocale,
```
Comment (compulim) [inline packages/bundle/src/createDirectLine.js:27]: Sort.

File: packages/embed/src/setups/version3.js [hunk 4]
```diff
@@ -17,7 +17,7 @@ export async function getBingSpeechToken(directLineToken, bingSpeechTokenURL) {
 export default async function setupVersionFamily3(
   { assets },
   { botId, directLineURL: domain, speechTokenURL, userId, webSocket },
-  { language, secret, token, username }
+  { language, secret, token, username, conversationStartLocale }
```
Comment (compulim) [inline packages/embed/src/setups/version3.js:20]: You don't need to touch this file. This is for Web Chat v3, which do not support the "Start Conversation" locale.

File: packages/embed/src/setups/version3.spec.js [hunk 5]
```diff
@@ -17,7 +17,8 @@ test('Load Web Chat v3 without a speech token', async () => {
           domain: 'https://directline.botframework.com',
           secret: 'secret',
           token: 'token',
-          webSocket: true
+          webSocket: true,
+          conversationStartLocale: 'en-US'
```
Comment (compulim) [inline packages/embed/src/setups/version3.spec.js:21]: This is for v3.

File: packages/embed/src/setups/version4.js [hunk 6]
```diff
@@ -4,11 +4,11 @@ import loadAsset from './loadAsset';
 export default async function setupVersionFamily4(
   { assets },
   { botIconURL, directLineURL: domain, userId, webSocket },
-  { language, secret, token, username }
+  { language, secret, token, username, conversationStartLocale }
```
Comment (compulim) [inline packages/embed/src/setups/version4.js:7]: This is more than just passing a parameter.

Today:

- We read the `&l=XXX` parameter, and then pass it to Web Chat UI (not DLJS)
   - Web Chat will:
      - Passthrough the locale to the bot (per activity), even if it is invalid
      - Normalize the language by guessing an invalid locale, e.g. `123` -> `en`, `zh-TW` -> `zh-Hant-TW`
      - Web Chat use the normalized version of UI locale
- Embed support two types of locale:
   - Azure locale, for example, `zh-hant.zh-hk` (this is using Web Chat inside Azure Portal)
      - When receiving `zh-hant.zh-hk`, the embed will pass `zh-Hant-HK` to Web Chat instead (ISO format)
      - The second part is the culture. For example, `zh-hant.zh-hk` will display UI in Traditional Chinese, and currency/datetime in Hong Kong format
   - ISO format locale, for example, `en` and `zh-Hant-HK`

Please ask @p-nagpal how to handle Azure locale:

1. Should we convert Azure locale to ISO locale before sending to the bot?
1. Should we passthrough Azure locale to the bot?

File: packages/embed/src/setups/version4.spec.js [hunk 7]
```diff
@@ -48,7 +48,8 @@ test('Load Web Chat v4', async () => {
       language: 'ja-JP',
       secret: 'secret',
       token: 'token',
-      username: 'William'
+      username: 'William',
+      conversationStartLocale: 'en-US'
```
Comment (compulim) [inline packages/embed/src/setups/version4.spec.js:52]: Sort.

File: packages/embed/src/setups/version4.js [hunk 8]
```diff
@@ -4,11 +4,11 @@ import loadAsset from './loadAsset';
...
 ) {
   assets && (await Promise.all(assets.map(loadAsset)));
 
-  const directLine = window.WebChat.createDirectLine({ domain, secret, token, webSocket });
+  const directLine = window.WebChat.createDirectLine({ domain, secret, token, webSocket, conversationStartLocale });
```
Comment (compulim) [inline packages/embed/src/setups/version4.js:11]: Sort.

File: packages/embed/src/setups/version4.spec.js [hunk 9]
```diff
@@ -65,7 +66,8 @@ test('Load Web Chat v4', async () => {
     domain: 'https://directline.botframework.com',
     secret: 'secret',
     token: 'token',
-    webSocket: true
+    webSocket: true,
+    conversationStartLocale: 'en-US'
```
Comment (compulim) [inline packages/embed/src/setups/version4.spec.js:70]: Sort.

**Comment by timenick** [comment]
zh-CN test case failed due to the locale feature is not enabled in DLJS.

File: __tests__/html/simple.noLocaleIsSent.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
+/**
+ * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
+ */
+
+describe('simple', () => {
+  test('should get "Hello and welcome!" message.', () => runHTMLTest('simple.noLocaleIsSent.html'));
```
Comment (compulim) [inline __tests__/html/simple.noLocaleIsSent.js:6]: Can you rename these tests?

For example, this one should be:

```js
describe('conversationStartProperties', () => {
  test('with no locale should get "Hello and welcome!" message.', () => runHTMLTest('simple.noLocaleIsSent.html'));
```

File: __tests__/html/conversationStartProperties.noLocaleIsSent.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+/**
+ * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
+ */
+
+describe('conversationStartProperties', () => {
+  test('should get "Hello and welcome!" message.', () =>
```
Comment (compulim) [inline __tests__/html/conversationStartProperties.noLocaleIsSent.js:6]:
```suggestion
  test('with no locale sent should get "Hello and welcome!" message.', () =>
```

File: __tests__/html/conversationStartProperties.sendEnUs.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+/**
+ * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
+ */
+
+describe('conversationStartProperties', () => {
+  test('should get "Hello and welcome!" message.', () => runHTMLTest('conversationStartProperties.sendEnUs.html'));
```
Comment (compulim) [inline __tests__/html/conversationStartProperties.sendEnUs.js:6]:
```suggestion
  test('with locale of en-US should get "Hello and welcome!" message.', () => runHTMLTest('conversationStartProperties.sendEnUs.html'));
```

File: __tests__/html/conversationStartProperties.sendInvalidType.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+/**
+ * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
+ */
+
+describe('conversationStartProperties', () => {
+  test('should get "Hello and welcome!" message.', () =>
```
Comment (compulim) [inline __tests__/html/conversationStartProperties.sendInvalidType.js:6]:
```suggestion
  test('with locale of invalid type should get "Hello and welcome!" message.', () =>
```

File: __tests__/html/conversationStartProperties.sendNonExisting.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+/**
+ * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
+ */
+
+describe('conversationStartProperties', () => {
+  test('should get "Hello and welcome!" message.', () =>
```
Comment (compulim) [inline __tests__/html/conversationStartProperties.sendNonExisting.js:6]:
```suggestion
  test('with non-existing locale should get "Hello and welcome!" message.', () =>
```

File: __tests__/html/conversationStartProperties.sendNonISOFormat.js [hunk 15]
```diff
@@ -0,0 +1,8 @@
+/**
+ * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
+ */
+
+describe('conversationStartProperties', () => {
+  test('should get "Hello and welcome!" message.', () =>
```
Comment (compulim) [inline __tests__/html/conversationStartProperties.sendNonISOFormat.js:6]:
```suggestion
  test('with non-ISO format locale should get "Hello and welcome!" message.', () =>
```

File: __tests__/html/conversationStartProperties.sendZhCn.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
+/**
+ * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
+ */
+
+describe('conversationStartProperties', () => {
+  test('should get "Hello and welcome!" in Simplified Chinese.', () =>
```
Comment (compulim) [inline __tests__/html/conversationStartProperties.sendZhCn.js:6]:
```suggestion
  test('with locale of zh-CN should get "Hello and welcome!" in Simplified Chinese.', () =>
```

**Comment by compulim** [comment]
LGTM!

**Pull request merged**